include/pybind11/pybind11.h

```diff
@@ -1374,6 +1374,141 @@ using default_holder_type = smart_holder;
         }
 
 #endif
+
+// Classic (non-smart_holder) implementations for .def_readonly and .def_readwrite
+// getter and setter functions.
+// WARNING: This classic implementation can lead to dangling pointers for raw pointer members.
+// See test_ptr() in tests/test_class_sh_property.py
+// This implementation works as-is for smart_holder std::shared_ptr members.
+template <typename T, typename D, typename SFINAE = void>
+struct xetter_cpp_function {
+    template <typename PM>
+    static cpp_function readonly(PM pm, const handle &hdl) {
+        return cpp_function([pm](const T &c) -> const D & { return c.*pm; }, is_method(hdl));
+    }
+
+    template <typename PM>
+    static cpp_function read(PM pm, const handle &hdl) {
+        return readonly(pm, hdl);
+    }
+
+    template <typename PM>
+    static cpp_function write(PM pm, const handle &hdl) {
+        return cpp_function([pm](T &c, const D &value) { c.*pm = value; }, is_method(hdl));
+    }
+};
+
+// smart_holder specializations for raw pointer members.
+// WARNING: Like the classic implementation, this implementation can lead to dangling pointers.
+// See test_ptr() in tests/test_class_sh_property.py
+// However, the read functions return a shared_ptr to the member, emulating PyCLIF approach:
+// https://github.com/google/clif/blob/c371a6d4b28d25d53a16e6d2a6d97305fb1be25a/clif/python/instance.h#L233
+// This prevents disowning of the Python object owning the raw pointer member.
+template <typename T, typename D>
+struct xetter_cpp_function<
+    T,
+    D,
+    detail::enable_if_t<detail::type_uses_smart_holder_type_caster<T>::value
+                        && detail::type_uses_smart_holder_type_caster<D>::value
+                        && std::is_pointer<D>::value>> {
+
+    using drp = typename std::remove_pointer<D>::type;
+
+    template <typename PM>
+    static cpp_function readonly(PM pm, const handle &hdl) {
```

Comment:
```suggestion
    static cpp_function readonly(PM &pm, const handle &hdl) {
```
and capture by reference in the lambda. At the very least, the PM should be forwarded otherwise.

I think original pybind11 it's a pointer. being passed around.

This could invoke copy or moves depending on the type PMs.

Reply:
I couldn't get this to work. I tried a few variations of sprinkling in `&`, `&&`, `std::forward<PM>`, but either it didn't build or I got segfaults. Then I decided it's better to just make it obvious that we're only dealing with pointers anyway, and the perfect forwarding doesn't buy us anything. See the long added comment for the new `must_be_member_function_pointer` SFINAE helper.